Cygwin CI job fails with `f2py` errors

See for example [this CI log](https://github.com/numpy/numpy/actions/runs/3981840176/jobs/6825839167). Failures:

 =========================== short test summary info ============================
FAILED ../../f2py/tests/test_compile_function.py::test_f2py_init_compile[extra_args0]
FAILED ../../f2py/tests/test_compile_function.py::test_compile_from_strings[program test_f2py\nend program test_f2py0]
FAILED ../../f2py/tests/test_compile_function.py::test_compile_from_strings[program test_f2py\nend program test_f2py1]
FAILED ../../tests/test_public_api.py::test_import_lazy_import[testing] - Blo...
ERROR ../../f2py/tests/test_character.py::TestMiscCharacter::test_gh18684 - B...
ERROR ../../f2py/tests/test_character.py::TestMiscCharacter::test_gh6308 - Bl...
ERROR ../../f2py/tests/test_character.py::TestMiscCharacter::test_gh4519 - Bl...
ERROR ../../f2py/tests/test_character.py::TestMiscCharacter::test_gh3425 - Bl...
ERROR ../../f2py/tests/test_character.py::TestMiscCharacter::test_character_bc[new]
ERROR ../../f2py/tests/test_character.py::TestMiscCharacter::test_character_bc[old]
ERROR ../../f2py/tests/test_crackfortran.py::TestCrackFortran::test_gh2848 - ...
ERROR ../../f2py/tests/test_string.py::TestDocStringArguments::test_example
= 4 failed, 27912 passed, 203 skipped, 1306 deselected, 35 xfailed, 1 xpassed, 8 errors in 546.40s (0:09:06) 

Cc @DWesl have you seen this before? May be due to the 22 Jan Cygwin release, not sure.

--- a/.github/workflows/cygwin.yml
+++ b/.github/workflows/cygwin.yml
@@ -79,7 +79,7 @@ jobs:
       - name: Run NumPy F2Py tests
         shell: "C:\\tools\\cygwin\\bin\\bash.exe -o igncr -eo pipefail {0}"
         # I need the separate processes more than I need the parallelization
-        run: /usr/bin/python3.9 runtests.py -n -s f2py -- --numprocesses=4;
+        run: /usr/bin/python3.9 runtests.py -n -s f2py -- --numprocesses=6;
       - name: Upload wheel if tests fail
         uses: actions/upload-artifact@v3
         if: failure()

--- a/.github/workflows/cygwin.yml
+++ b/.github/workflows/cygwin.yml
@@ -79,7 +79,7 @@ jobs:
       - name: Run NumPy F2Py tests
         shell: "C:\\tools\\cygwin\\bin\\bash.exe -o igncr -eo pipefail {0}"
         # I need the separate processes more than I need the parallelization
-        run: /usr/bin/python3.9 runtests.py -n -s f2py -- --numprocesses=6;
+        run: /usr/bin/python3.9 runtests.py -n -s f2py -- --numprocesses=auto;
       - name: Upload wheel if tests fail
         uses: actions/upload-artifact@v3
         if: failure()

--- a/.github/workflows/cygwin.yml
+++ b/.github/workflows/cygwin.yml
@@ -56,7 +56,7 @@ jobs:
           dash -c "which python3.9; /usr/bin/python3.9 --version -V"
       - name: Build NumPy wheel
         run: |
-          dash -c "/usr/bin/python3.9 -m pip install 'setuptools<49.2.0' pytest pytz cffi pickle5 importlib_metadata typing_extensions pytest-xdist"
+          dash -c "/usr/bin/python3.9 -m pip install 'setuptools<49.2.0' pytest pytz cffi pickle5 importlib_metadata typing_extensions"
           dash -c "/usr/bin/python3.9 -m pip install -r test_requirements.txt"
           dash -c "/usr/bin/python3.9 setup.py bdist_wheel"
       - name: Install new NumPy
@@ -71,15 +71,7 @@ jobs:
         # Not sure if that will run the top-level tests.
         shell: "C:\\tools\\cygwin\\bin\\bash.exe -o igncr -eo pipefail {0}"
         run: |
-          for submodule in array_api compat core distutils fft lib linalg ma matrixlib polynomial random tests typing;
-          do 
-              echo "Testing numpy submodule ${submodule}";
-              /usr/bin/python3.9 runtests.py -n -s "${submodule}" || exit 1; 
-          done;
-      - name: Run NumPy F2Py tests
-        shell: "C:\\tools\\cygwin\\bin\\bash.exe -o igncr -eo pipefail {0}"
-        # I need the separate processes more than I need the parallelization
-        run: /usr/bin/python3.9 runtests.py -n -s f2py -- --numprocesses=auto;
+          /usr/bin/python3.9 runtests.py -n
       - name: Upload wheel if tests fail
         uses: actions/upload-artifact@v3
         if: failure()

--- a/.github/workflows/cygwin.yml
+++ b/.github/workflows/cygwin.yml
@@ -56,7 +56,7 @@ jobs:
           dash -c "which python3.9; /usr/bin/python3.9 --version -V"
       - name: Build NumPy wheel
         run: |
-          dash -c "/usr/bin/python3.9 -m pip install 'setuptools<49.2.0' pytest pytz cffi pickle5 importlib_metadata typing_extensions pytest-forked"
+          dash -c "/usr/bin/python3.9 -m pip install 'setuptools<49.2.0' pytest pytz cffi pickle5 importlib_metadata typing_extensions pytest-xdist"
           dash -c "/usr/bin/python3.9 -m pip install -r test_requirements.txt"
           dash -c "/usr/bin/python3.9 setup.py bdist_wheel"
       - name: Install new NumPy
@@ -78,7 +78,8 @@ jobs:
           done;
       - name: Run NumPy F2Py tests
         shell: "C:\\tools\\cygwin\\bin\\bash.exe -o igncr -eo pipefail {0}"
-        run: /usr/bin/python3.9 runtests.py -n -s f2py -- --forked;
+        # I need the separate processes more than I need the parallelization
+        run: /usr/bin/python3.9 runtests.py -n -s f2py -- --numprocesses=4;
       - name: Upload wheel if tests fail
         uses: actions/upload-artifact@v3
         if: failure()

--- a/.github/workflows/cygwin.yml
+++ b/.github/workflows/cygwin.yml
@@ -65,19 +65,20 @@ jobs:
       - name: Rebase NumPy compiled extensions
         run: |
           dash "tools/rebase_installed_dlls_cygwin.sh" 3.9
-      - name: Run NumPy test suite
+      - name: Run NumPy test suite except F2PY
         # There seem to be too many compiled extension modules.
         # Let's try splitting off the tests by submodule to see if that helps.
         # Not sure if that will run the top-level tests.
-        shell: "C:\\tools\\cygwin\\bin\\bash.exe -o igncr -o pipefail {0}"
-        run: >-
+        shell: "C:\\tools\\cygwin\\bin\\bash.exe -o igncr -eo pipefail {0}"
+        run: |
           for submodule in array_api compat core distutils fft lib linalg ma matrixlib polynomial random tests typing;
           do 
               echo "Testing numpy submodule ${submodule}";
               /usr/bin/python3.9 runtests.py -n -s "${submodule}" || exit 1; 
-          done
-          echo "Testing numpy submodule f2py";
-          /usr/bin/python3.9 runtests.py -n -s f2py -- --forked
+          done;
+      - name: Run NumPy F2Py tests
+        shell: "C:\\tools\\cygwin\\bin\\bash.exe -o igncr -eo pipefail {0}"
+        run: /usr/bin/python3.9 runtests.py -n -s f2py -- --forked;
       - name: Upload wheel if tests fail
         uses: actions/upload-artifact@v3
         if: failure()

--- a/.github/workflows/cygwin.yml
+++ b/.github/workflows/cygwin.yml
@@ -65,10 +65,7 @@ jobs:
       - name: Rebase NumPy compiled extensions
         run: |
           dash "tools/rebase_installed_dlls_cygwin.sh" 3.9
-      - name: Run NumPy test suite except F2PY
-        # There seem to be too many compiled extension modules.
-        # Let's try splitting off the tests by submodule to see if that helps.
-        # Not sure if that will run the top-level tests.
+      - name: Run NumPy test suite
         shell: "C:\\tools\\cygwin\\bin\\bash.exe -o igncr -eo pipefail {0}"
         run: |
           /usr/bin/python3.9 runtests.py -n

--- a/.github/workflows/cygwin.yml
+++ b/.github/workflows/cygwin.yml
@@ -56,7 +56,7 @@ jobs:
           dash -c "which python3.9; /usr/bin/python3.9 --version -V"
       - name: Build NumPy wheel
         run: |
-          dash -c "/usr/bin/python3.9 -m pip install 'setuptools<49.2.0' pytest pytz cffi pickle5 importlib_metadata typing_extensions"
+          dash -c "/usr/bin/python3.9 -m pip install 'setuptools<49.2.0' pytest pytz cffi pickle5 importlib_metadata typing_extensions pytest-forked"
           dash -c "/usr/bin/python3.9 -m pip install -r test_requirements.txt"
           dash -c "/usr/bin/python3.9 setup.py bdist_wheel"
       - name: Install new NumPy
@@ -65,17 +65,19 @@ jobs:
       - name: Rebase NumPy compiled extensions
         run: |
           dash "tools/rebase_installed_dlls_cygwin.sh" 3.9
-      - name: Run NumPy test suite except F2PY
+      - name: Run NumPy test suite
         # There seem to be too many compiled extension modules.
         # Let's try splitting off the tests by submodule to see if that helps.
         # Not sure if that will run the top-level tests.
         shell: "C:\\tools\\cygwin\\bin\\bash.exe -o igncr -o pipefail {0}"
         run: >-
-          for submodule in array_api compat core distutils f2py fft lib linalg ma matrixlib polynomial random tests typing;
+          for submodule in array_api compat core distutils fft lib linalg ma matrixlib polynomial random tests typing;
           do 
               echo "Testing numpy submodule ${submodule}";
               /usr/bin/python3.9 runtests.py -n -s "${submodule}" || exit 1; 
           done
+          echo "Testing numpy submodule f2py";
+          /usr/bin/python3.9 runtests.py -n -s f2py -- --forked
       - name: Upload wheel if tests fail
         uses: actions/upload-artifact@v3
         if: failure()

--- a/.github/workflows/cygwin.yml
+++ b/.github/workflows/cygwin.yml
@@ -67,12 +67,14 @@ jobs:
           dash "tools/rebase_installed_dlls_cygwin.sh" 3.9
       - name: Run NumPy test suite except F2PY
         # There seem to be too many compiled extension modules.
-        # Let's try splitting off the F2PY tests to see if that helps
+        # Let's try splitting off the tests by submodule to see if that helps.
+        # Not sure if that will run the top-level tests.
+        shell: "C:\\tools\\cygwin\\bin\\bash.exe -o igncr -o pipefail {0}"
         run: >-
-          dash -c "/usr/bin/python3.9 runtests.py -n -- -k 'not f2py'"
-      - name: Run NumPy F2PY tests
-        run: >-
-          dash -c "/usr/bin/python3.9 runtests.py -n -- -k 'f2py'"
+          for submodule in array_api compat core distutils f2py fft lib linalg ma matrixlib polynomial random tests typing;
+          do 
+              /usr/bin/python3.9 runtests.py -n -s "${submodule}"; 
+          done
       - name: Upload wheel if tests fail
         uses: actions/upload-artifact@v3
         if: failure()

--- a/.github/workflows/cygwin.yml
+++ b/.github/workflows/cygwin.yml
@@ -73,7 +73,8 @@ jobs:
         run: >-
           for submodule in array_api compat core distutils f2py fft lib linalg ma matrixlib polynomial random tests typing;
           do 
-              /usr/bin/python3.9 runtests.py -n -s "${submodule}"; 
+              echo "Testing numpy submodule ${submodule}";
+              /usr/bin/python3.9 runtests.py -n -s "${submodule}" || exit 1; 
           done
       - name: Upload wheel if tests fail
         uses: actions/upload-artifact@v3

--- a/.github/workflows/cygwin.yml
+++ b/.github/workflows/cygwin.yml
@@ -65,9 +65,14 @@ jobs:
       - name: Rebase NumPy compiled extensions
         run: |
           dash "tools/rebase_installed_dlls_cygwin.sh" 3.9
-      - name: Run NumPy test suite
+      - name: Run NumPy test suite except F2PY
+        # There seem to be too many compiled extension modules.
+        # Let's try splitting off the F2PY tests to see if that helps
         run: >-
-          dash -c "/usr/bin/python3.9 runtests.py -n"
+          dash -c "/usr/bin/python3.9 runtests.py -n -- --ignore=numpy/f2py"
+      - name: Run NumPy F2PY tests
+        run: >-
+          dash -c "/usr/bin/python3.9 runtests.py -n -s f2py
       - name: Upload wheel if tests fail
         uses: actions/upload-artifact@v3
         if: failure()

--- a/.github/workflows/cygwin.yml
+++ b/.github/workflows/cygwin.yml
@@ -69,10 +69,10 @@ jobs:
         # There seem to be too many compiled extension modules.
         # Let's try splitting off the F2PY tests to see if that helps
         run: >-
-          dash -c "/usr/bin/python3.9 runtests.py -n -- --ignore=numpy/f2py"
+          dash -c "/usr/bin/python3.9 runtests.py -n -- -k 'not f2py'"
       - name: Run NumPy F2PY tests
         run: >-
-          dash -c "/usr/bin/python3.9 runtests.py -n -s f2py
+          dash -c "/usr/bin/python3.9 runtests.py -n -- -k 'f2py'"
       - name: Upload wheel if tests fail
         uses: actions/upload-artifact@v3
         if: failure()

--- a/numpy/f2py/tests/util.py
+++ b/numpy/f2py/tests/util.py
@@ -30,6 +30,9 @@ from importlib import import_module
 _module_dir = None
 _module_num = 5403
 
+if sys.platform == "cygwin":
+    _module_list = []
+
 
 def _cleanup():
     global _module_dir
@@ -146,6 +149,19 @@ def build_module(source_files, options=[], skip=[], only=[], module_name=None):
         # Partial cleanup
         for fn in dst_sources:
             os.unlink(fn)
+
+    # Rebase
+    if sys.platform == "cygwin":
+        # If someone starts deleting modules after import, this will
+        # need to change to record how big each module is, rather than
+        # relying on rebase being able to find that from the files.
+        _module_list.extend(
+            glob.glob(os.path.join(d, "{:s}*".format(module_name)))
+        )
+        subprocess.check_call(
+            ["/usr/bin/rebase", "--database", "--oblivious", "--verbose"]
+            + _module_list
+        )
 
     # Import
     return import_module(module_name)

--- a/numpy/f2py/tests/util.py
+++ b/numpy/f2py/tests/util.py
@@ -14,7 +14,6 @@ import shutil
 import atexit
 import textwrap
 import re
-import glob
 import pytest
 import contextlib
 import numpy

--- a/numpy/f2py/tests/util.py
+++ b/numpy/f2py/tests/util.py
@@ -6,6 +6,7 @@ Utility functions for
 - determining paths to tests
 
 """
+import glob
 import os
 import sys
 import subprocess
@@ -29,6 +30,9 @@ from importlib import import_module
 
 _module_dir = None
 _module_num = 5403
+
+if sys.platform == "cygwin":
+    _module_list = []
 
 
 def _cleanup():
@@ -146,6 +150,21 @@ def build_module(source_files, options=[], skip=[], only=[], module_name=None):
         # Partial cleanup
         for fn in dst_sources:
             os.unlink(fn)
+
+    # Rebase (Cygwin-only)
+    if sys.platform == "cygwin":
+        # If someone starts deleting modules after import, this will
+        # need to change to record how big each module is, rather than
+        # relying on rebase being able to find that from the files.
+        _module_list.extend(
+            glob.glob(os.path.join(d, "{:s}*".format(module_name)))
+        )
+        subprocess.check_call(
+            ["/usr/bin/rebase", "--database", "--oblivious", "--verbose"]
+            + _module_list
+        )
+
+
 
     # Import
     return import_module(module_name)

--- a/numpy/f2py/tests/util.py
+++ b/numpy/f2py/tests/util.py
@@ -31,9 +31,6 @@ from importlib import import_module
 _module_dir = None
 _module_num = 5403
 
-if sys.platform == "cygwin":
-    _module_list = []
-
 
 def _cleanup():
     global _module_dir
@@ -150,19 +147,6 @@ def build_module(source_files, options=[], skip=[], only=[], module_name=None):
         # Partial cleanup
         for fn in dst_sources:
             os.unlink(fn)
-
-    # Rebase
-    if sys.platform == "cygwin":
-        # If someone starts deleting modules after import, this will
-        # need to change to record how big each module is, rather than
-        # relying on rebase being able to find that from the files.
-        _module_list.extend(
-            glob.glob(os.path.join(d, "{:s}*".format(module_name)))
-        )
-        subprocess.check_call(
-            ["/usr/bin/rebase", "--database", "--oblivious", "--verbose"]
-            + _module_list
-        )
 
     # Import
     return import_module(module_name)

--- a/numpy/f2py/tests/util.py
+++ b/numpy/f2py/tests/util.py
@@ -32,7 +32,8 @@ _module_dir = None
 _module_num = 5403
 
 if sys.platform == "cygwin":
-    _module_list = []
+    NUMPY_INSTALL_ROOT = Path(__file__).parent.parent.parent
+    _module_list = list(NUMPY_INSTALL_ROOT.glob("**/*.dll"))
 
 
 def _cleanup():

--- a/numpy/f2py/tests/util.py
+++ b/numpy/f2py/tests/util.py
@@ -14,6 +14,7 @@ import shutil
 import atexit
 import textwrap
 import re
+import glob
 import pytest
 import contextlib
 import numpy

--- a/runtests.py
+++ b/runtests.py
@@ -545,6 +545,17 @@ def build_project(args):
             print("Build failed!")
         sys.exit(1)
 
+    # Rebase
+    if sys.platform == "cygwin":
+        from pathlib import path
+        testenv_root = Path(config_vars["platbase"])
+        dll_list = testenv_root.glob("**/*.dll")
+        rebase_cmd = ["/usr/bin/rebase", "--database", "--oblivious"]
+        rebase_cmd.extend(dll_list)
+        if subprocess.run(rebase_cmd):
+            print("Rebase failed")
+            sys.exit(1)
+
     return site_dir, site_dir_noarch
 
 def asv_compare_config(bench_path, args, h_commits):

--- a/tools/list_installed_dll_dependencies_cygwin.sh
+++ b/tools/list_installed_dll_dependencies_cygwin.sh
@@ -14,11 +14,11 @@
 py_ver=${1}
 dll_list=`/bin/dash tools/list_numpy_dlls.sh ${py_ver}`
 echo "Checks for existence, permissions and file type"
-ls -l ${dll_list}
-file ${dll_list}
+/usr/bin/timeout 10m /usr/bin/ls -l ${dll_list}
+/usr/bin/timeout 10m /usr/bin/file ${dll_list}
 echo "Dependency checks"
-ldd ${dll_list} | grep -F -e " => not found" && exit 1
-cygcheck ${dll_list} >cygcheck_dll_list 2>cygcheck_missing_deps
+/usr/bin/timeout 10m /usr/bin/ldd ${dll_list} | grep -F -e " => not found" && exit 1
+/usr/bin/timeout 10m /usr/bin/cygcheck ${dll_list} >cygcheck_dll_list 2>cygcheck_missing_deps
 grep -F -e "cygcheck: track_down: could not find " cygcheck_missing_deps && exit 1
 echo "Import tests"
 mkdir -p dist/
@@ -31,5 +31,5 @@ do
 			 -e "s/^\/+(home|usr).*?site-packages\/+//" \
 			 -e "s/.cpython-3.m?-x86(_64)?-cygwin.dll$//" \
 			 -e "s/\//./g"`
-    python${py_ver} -c "import ${ext_module}"
+    /usr/bin/timeout 2m /usr/bin/python${py_ver} -c "import ${ext_module}"
 done

--- a/tools/rebase_installed_dlls_cygwin.sh
+++ b/tools/rebase_installed_dlls_cygwin.sh
@@ -2,4 +2,6 @@
 # Rebase the dlls installed by NumPy
 
 py_ver=${1}
-/usr/bin/rebase --database --oblivious `/bin/dash tools/list_numpy_dlls.sh ${py_ver}`
+numpy_dlls="`/bin/dash tools/list_numpy_dlls.sh ${py_ver}`"
+/usr/bin/rebase --verbose --database --oblivious ${numpy_dlls}
+/usr/bin/rebase --verbose --info ${numpy_dlls}

--- a/tools/rebase_installed_dlls_cygwin.sh
+++ b/tools/rebase_installed_dlls_cygwin.sh
@@ -3,5 +3,6 @@
 
 py_ver=${1}
 numpy_dlls="`/bin/dash tools/list_numpy_dlls.sh ${py_ver}`"
-/usr/bin/rebase --verbose --database --oblivious ${numpy_dlls}
+# Add --oblivious to next line if not run as root
+/usr/bin/rebase --verbose --database ${numpy_dlls}
 /usr/bin/rebase --verbose --info ${numpy_dlls}

--- a/tools/rebase_installed_dlls_cygwin.sh
+++ b/tools/rebase_installed_dlls_cygwin.sh
@@ -3,6 +3,5 @@
 
 py_ver=${1}
 numpy_dlls="`/bin/dash tools/list_numpy_dlls.sh ${py_ver}`"
-# Add --oblivious to next line if not run as root
-/usr/bin/rebase --verbose --database ${numpy_dlls}
+/usr/bin/rebase --verbose --database --oblivious ${numpy_dlls}
 /usr/bin/rebase --verbose --info ${numpy_dlls}

